File: Dynamic_Programming/1473. Paint House III.cpp
/*
1473. Paint House III
*/

// 優化 TC:O(m^2*n) SC:O(m*n*k)
class Solution {
    int dp[105][105][25];
public:
    int minCost(vector<int>& houses, vector<vector<int>>& cost, int m, int n, int target) {
        //遍歷房子
        houses.insert(houses.begin(), 0);
        cost.insert(cost.begin(), {0});
        for(int i = 0; i<= m; i++){
            for(int j = 0; j<= m; j++){
                for(int k = 0; k <= n; k++){
                    dp[i][j][k] = INT_MAX/2;
                }
            }
        }
        //初始值 如果第一個房子被粉過就給他0 沒有的話去找要粉他的價格
        if (houses[1]!=0){
            dp[1][1][houses[1]] = 0;
        }
        else {
            for (int k=1; k<=n; k++)
                dp[1][1][k] = cost[1][k-1];
        }

        for(int i = 2; i <= m; i++){
        //遍歷造成了幾個neighbor -> block
            //已經確認了要這個顏色
            if(houses[i] != 0){
                for(int j = 1; j <= target; j++){
                    //遍歷選擇的顏色
                    int k = houses[i];
                    for(int kk = 1; kk <= n; kk++){
                        if(kk != k){
                            dp[i][j][k] = min(dp[i][j][k] ,dp[i-1][j-1][kk]);
                        }
                        else{
                            dp[i][j][k] = min(dp[i][j][k],dp[i-1][j][kk]);
                        }
                    }
                }
            }
            else {
                for(int j = 1; j <= target; j++){
                    vector<pair<int,int>> temp;//最小值 對應顏色
                    //先將之前狀態的每個顏色最小值抓出 就不用重複遍歷了
                    for(int kk = 1; kk<=n; kk++){
                        temp.push_back({dp[i-1][j-1][kk], kk});
                    }
                    sort(temp.begin(),temp.end());

                    for(int k = 1; k<=n; k++){
                        //先將一樣的狀態加進去
                        dp[i][j][k] = dp[i-1][j][k] + cost[i][k-1];

                        if(temp[0].second != k){
                            dp[i][j][k] = min(dp[i][j][k] , temp[0].first + cost[i][k-1]);
                        }
                        else{
                            dp[i][j][k] = min(dp[i][j][k], temp[1].first + cost[i][k-1]);
                        }
                    
                    }
                }
            }
        }

        int res = INT_MAX/2;
        for(int k = 1; k<=n; k++){
            res = min(res, dp[m][target][k]);
        }
        if(res == INT_MAX/2) return -1;
        else return res; 
    }
};


// TC:O(m^2*n^2) SC:O(m*n*k)
class Solution {
    int dp[105][105][25];
public:
    int minCost(vector<int>& houses, vector<vector<int>>& cost, int m, int n, int target) {
        //遍歷房子
        houses.insert(houses.begin(), 0);
        cost.insert(cost.begin(), {0});
        for(int i = 0; i<= m; i++){
            for(int j = 0; j<= m; j++){
                for(int k = 0; k <= n; k++){
                    dp[i][j][k] = INT_MAX/2;
                }
            }
        }
        //初始值
        if (houses[1]!=0)
        {
            dp[1][1][houses[1]] = 0;
        }
        else
        {
            for (int k=1; k<=n; k++)
                dp[1][1][k] = cost[1][k-1];
        }

        for(int i = 2; i <= m; i++){
        //遍歷造成了幾個neighbor -> block
            //已經確認了要這個顏色
            if(houses[i] != 0){
                for(int j = 1; j <= target; j++){
                    //遍歷選擇的顏色
                    int k = houses[i];
                    for(int kk = 1; kk <= n; kk++){
                        if(kk != k){
                            dp[i][j][k] = min(dp[i][j][k] ,dp[i-1][j-1][kk]);
                        }
                        else{
                            dp[i][j][k] = min(dp[i][j][k],dp[i-1][j][kk]);
                        }
                    }
                }
            }
            else {
                for(int j = 1; j <= target; j++){
                    for(int k = 1; k<=n; k++){
                        for(int kk = 1; kk<=n; kk++){
                            if(kk != k){
                                dp[i][j][k] = min(dp[i][j][k] ,dp[i-1][j-1][kk] + cost[i][k-1]);
                            }
                            else{
                                dp[i][j][k] = min(dp[i][j][k],dp[i-1][j][kk] + cost[i][k-1]);
                            }
                        }
                    }
                }
            }
        }

        int res = INT_MAX/2;
        for(int k = 1; k<=n; k++){
            res = min(res, dp[m][target][k]);
        }
        if(res == INT_MAX/2) return -1;
        else return res; 
    }
};


/*
此題突破口在於說限制了neigbor種類要到target 所以會想到每次塗的顏色會跟前面有關 無後效性->dp
想找最低cost達成目標
//遍歷房子
for(int i = 1; i <= m; i++){
//遍歷造成了幾個neighbor -> block
    //已經確認了要這個顏色
    if(house[i] != 0){
        for(int j = 1; j <= target; j++){
            //遍歷選擇的顏色
            int k = house[i];
            for(int kk = 1; kk <= n; kk++){
                if(kk != k){
                    dp[i][j][k] = min(dp[i][j][kk] ,dp[i-1][j-1][kk]);
                }
                else{
                    dp[i][j][k] = min(dp[i][j][k],dp[i-1][j][kk]);
                }
            }
        }
    }
    else {
        for(int j = 1; j <= target; j++){
            for(int k = 1; k<=n; k++){
                for(int kk = 1; kk<=n; kk++){
                    if(kk != k){
                        dp[i][j][k] = min(dp[i][j][kk] ,dp[i-1][j-1][kk] + cost[i][k]);
                    }
                    else{
                        dp[i][j][k] = min(dp[i][j][k],dp[i-1][j][kk] + cost[i][k]);
                    }
                }
            }
        }
    }
}
*/

File: Dynamic_Programming/1473. Paint House III.py
# TC:O(m * m * n * target) ->  TC:O(m * n * target) SC:O(m*n*target)
class Solution:
    def minCost(self, houses: List[int], cost: List[List[int]], m: int, n: int, target: int) -> int:
        f = [[[inf] * (n+1) for _ in range(target + 1)] for _ in range(m+1)]
        f[0][0][0] = 0
        # 時間優化 
        for i in range(m): # 走到哪棟房子
            if houses[i] != 0:
                for j in range(0, min(target,i+1) + 1): # 前面可能有幾組 + 我這組
                    for k in range(n + 1):
                        if f[i][j][k] != inf:
                            new_group = j + (1 if houses[i] != k else 0)
                            if new_group <= target:
                                f[i+1][new_group][houses[i]] = min(f[i+1][new_group][houses[i]], f[i][j][k])
            else:
                for j in range(0, min(target,i+1) + 1): # 前面可能有幾組 + 我這組 or 不加我這組
                    mn1, mn2, c1, c2 = inf, inf, 0, 0
                    # 後面再做的就是從之前不同的顏色中選最小跟次小的出來 :)
                    for t in range(0, n + 1): # 現在選的顏色跟之前一樣
                        if t != 0:
                            f[i+1][j][t] = min(f[i+1][j][t], f[i][j][t] + cost[i][t-1])
                        if f[i][j][t] <= mn1:
                            mn2, mn1 = mn1, f[i][j][t]
                            c1, c2 = t, c1
                        elif f[i][j][t] < mn2:
                            c2 = t
                            mn2 = f[i][j][t]
                    if j + 1 <= target:
                        for t in range(1, n+1):
                            if t != c1: f[i+1][j+1][t] = mn1 + cost[i][t-1] if mn1 + cost[i][t-1] < f[i+1][j+1][t] else f[i+1][j+1][t]
                            else: f[i+1][j+1][t] = mn2 + cost[i][t-1] if mn2 + cost[i][t-1] < f[i+1][j+1][t] else f[i+1][j+1][t]
        ans = min(f[m][target][i] for i in range(1, n+1))
        return ans if ans != inf else -1

        '''
        遞推
        for i in range(m): # 走到哪棟房子
            if houses[i] != 0:
                for j in range(0, min(target,i+1) + 1): # 前面可能有幾組 + 我這組
                    for k in range(n + 1):
                        if f[i][j][k] != inf:
                            new_group = j + (1 if houses[i] != k else 0)
                            if new_group <= target:
                                f[i+1][new_group][houses[i]] = min(f[i+1][new_group][houses[i]], f[i][j][k])
            else:
                for j in range(0, min(target,i+1) + 1): # 前面可能有幾組 + 我這組
                    for k in range(n + 1):
                        if f[i][j][k] == inf: continue
                        for t in range(1, n + 1):
                            new_group = j + (1 if k != t else 0)
                            if new_group <= target:
                                f[i+1][new_group][t] = min(f[i+1][new_group][t], f[i][j][k] + cost[i][t-1])
        ans = min(f[m][target][i] for i in range(1, n+1))
        return ans if ans != inf else -1
        '''
        '''
        遞歸
        @cache
        def dfs(i:int, j:int, k:int): # painting house, groups, prev color
            if i < 0 and j == target:
                return 0
            if j > target or i < 0:
                return inf
            ans = inf
            if houses[i] != 0:
                ans = dfs(i-1, j + (1 if houses[i] != k else 0), houses[i])
            else:
                for t in range(n):
                    ans = min(ans, dfs(i-1, j + (1 if t+1 != k else 0), t+1) + cost[i][t])
            return ans
        ret = dfs(m-1, 0, 0)
        return ret if ret != inf else -1
        '''
